src/unfold.rs

```diff
@@ -0,0 +1,167 @@
+        initial: Result<(Option<Link>, Source), Error>,
+        to_items: Box<dyn Fn(Source) -> Vec<StreamOk> + Send + Sync>,
+    ) -> Self {
+        let dummy = Self {
+            github,
+            to_items,
+            items: None,
+            next_page: None,
+        };
+
+        dummy.load_state(initial)
+    }
+
+    async fn next(self) -> Option<(Result<StreamOk, Error>, Self)> {
+        let mut state = self;
+        loop {
+            match state.flat_next() {
+                FetcherState::NextReady(item, state) => {
+                    return Some((item, state));
+                }
+                FetcherState::Empty => {
+                    return None;
+                }
+                FetcherState::FetchNextPage(next_state) => {
+                    state = next_state.fetch_next_page().await;
+                }
+            }
+        }
+    }
+
+    fn flat_next(mut self) -> FetcherState<StreamOk, Self> {
+        if let Some(resultitems) = self.items.take() {
+            match resultitems {
+                Ok(mut items) => {
+                    if let Some(item) = items.next() {
+                        FetcherState::NextReady(
+                            Ok(item),
+                            Self {
+                                github: self.github,
+                                to_items: self.to_items,
+                                items: Some(Ok(items)),
+                                next_page: self.next_page,
+                            },
+                        )
+                    } else if self.next_page.is_some() {
+                        FetcherState::FetchNextPage(self)
+                    } else {
+                        // No more things, no more next pages
+                        FetcherState::Empty
+                    }
+                }
+                Err(e) => {
+                    FetcherState::NextReady(
+                        Err(e),
+                        Self {
+                            github: self.github,
+                            to_items: self.to_items,
+                            items: None,     // Returned once, just now
+                            next_page: None, // Cannot have a next_page if Items is Err !!! fix the data model
+                        },
+                    )
+                }
+            }
+        } else {
+            if self.next_page.is_some() {
+                FetcherState::FetchNextPage(self)
+            } else {
+                // No more things, no more next pages
+                FetcherState::Empty
+            }
+        }
+    }
+
+    async fn fetch_next_page(mut self) -> Self {
+        if let Some(next_page) = self.next_page.take() {
+            if let Some(url) = next_link(&next_page) {
+                if let Ok(url) = Url::parse(&url) {
+                    let uri = [url.path(), url.query().unwrap_or_default()].join("?");
+
+                    let fetched_page = self.github.get_pages(&uri).await;
```

Comment:
```suggestion
                    let fetched_page = self.github.get_pages(url.as_str()).await;
```
(This passes the unit tests, I don't know if the URLs are correct this way.)
Better yet would be for `get_pages` to take an argument of type `Url` directly.